src/digest/base.cr

```diff
@@ -1,92 +1,158 @@
 require "base64"
 
+class Digest::FinalizedError < Exception
+end
+
 abstract class Digest::Base
-  # Returns the hash of *data*. *data* must respond to `#to_slice`.
-  def self.digest(data)
-    digest do |ctx|
-      ctx.update(data.to_slice)
+  macro inherited
+    # Returns the hash of *data*. *data* must respond to `#to_slice`.
+    def self.digest(data)
+      digest do |ctx|
+        ctx.update(data.to_slice)
+      end
+    end
+
+    # Yields an instance of `self` which can receive calls to `#update(data : String | Bytes)` and returns the finalized digest.
+    # method available. Returns the resulting digest afterwards.
+    #
+    # ```
+    # require "digest/md5"
+    #
+    # digest = Digest::MD5.digest do |ctx|
+    #   ctx.update "f"
+    #   ctx.update "oo"
+    # end
+    # digest.to_slice.hexstring # => "acbd18db4cc2f85cedef654fccc4a4d8"
+    # ```
+    def self.digest(& : Digest::Base -> _) : Bytes
+      context = new
+      yield context
+      context.final
+    end
+
+    # Returns the hexadecimal representation of the hash of *data*.
+    #
+    # ```
+    # require "digest/md5"
+    #
+    # Digest::MD5.hexdigest("foo") # => "acbd18db4cc2f85cedef654fccc4a4d8"
+    # ```
+    def self.hexdigest(data) : String
+      hexdigest &.update(data)
+    end
+
+    # Yields a context object with an `#update(data : String | Bytes)`
+    # method available. Returns the resulting digest in hexadecimal representation
+    # afterwards.
+    #
+    # ```
+    # require "digest/md5"
+    #
+    # Digest::MD5.hexdigest("foo") # => "acbd18db4cc2f85cedef654fccc4a4d8"
+    # Digest::MD5.hexdigest do |ctx|
+    #   ctx.update "f"
+    #   ctx.update "oo"
+    # end
+    # # => "acbd18db4cc2f85cedef654fccc4a4d8"
+    # ```
+    def self.hexdigest(& : Digest::Base -> _) : String
+      hashsum = digest do |ctx|
+        yield ctx
+      end
+
+      hashsum.to_slice.hexstring
+    end
+
+    # Returns the base64-encoded hash of *data*.
+    #
+    # ```
+    # require "digest/sha1"
+    #
+    # Digest::SHA1.base64digest("foo") # => "C+7Hteo/D9vJXQ3UfzxbwnXaijM="
+    # ```
+    def self.base64digest(data) : String
+      base64digest &.update(data)
+    end
+
+    # Yields a context object with an `#update(data : String | Bytes)`
+    # method available. Returns the resulting digest in base64 representation
```

Comment:
This is inconsistent with `self.digest` above: either mention a context object or `self`, but not both wordings.

Reply:
I didn't write that comment.  The format tool reindented it.  It's already committed in crystal.

Is fixing other peoples comments on methods I haven't changed the function of related to this PR?  When I wrote the title I meant it to be a minimal cleanup, not taking on every task.  

Am I supposed to fix everything in `Digest` I didn't write before I get my PR through or is this PR already large/complicated and takes a long time to review? 

This seems like I'm being told conflicting information about the review process.    What will get my PR through faster?

Reply:
@didactic-drunk Please stay calm. There's no reason to assume that @RX14 was aware that this doc comment is not related to the change. It looks like it does if you don't ignore whitespace changes.
I understand you might be frustrated about not getting merged already, but your reaction is out of line. A simple note would've been enough.

Reply:
> There's no reason to assume that @RX14 was aware that this doc comment is not related to the change. It looks like it does if you don't ignore whitespace changes.

@straight-shoota If someone is undertaking a task of doing a code-review, yes, I'd say that should be **the reason** to assume he understands the changes he's looking at.

> I understand you might be frustrated about not getting merged already, but your reaction is out of line. A simple note would've been enough.

IMO it's a perfectly understandable reaction (frustration as you've noted) to the situation going on with this - and other PRs - by @didactic-drunk. People have emotions, which are repeatedly put to the tests by the actions of the Core Team members - like lack of communication, stalling PRs and others. I don't think this reaction is out of line, more like insinuating that after ~half of year of waiting, someones irritation might be misplaced. Above all, all he gets from you, is the _understanding_ of the faulty, or lack of the processes on your side that leads to situation like this - perhaps it's a good time to implement some bot to remind you about PRs, or take other corrective actions instead of _sympathizing_...

Reply:
I'm sorry that everyone has limited resources. And it seems there are not many people interested in reviewing this. I don't know why. With that I don't just mean core team members. There are no approving reviews of non-core members either which would certainly help to push this.

But I also don't think the tone of @didactic-drunk's previous comment is appropriate. And sadly I'm sure that such an attutide doesn't help to encourage anyone to do reviews.

Reply:
You changed the doc comment above:

![](https://elixi.re/i/wbphkymx.png)

I'm pointing out that if you change one doc comment, you need to change them all, or change none of them.

Reply:
> I'm sorry that everyone has limited resources. And it seems there are not many people interested in reviewing this. I don't know why. With that I don't just mean core team members.

No need to be sorry, it ain't about limited resources, but the communication - or lack thereof - which is a major hurdle for everyone involved. It would be way easier if we could have any estimates in re: to when can we expect any reaction, who's responsible for doing so, what can we do to ease the process as much as possible (thinking of you), etc. - and all of that we must find out by ourselves, since there's no formal review process, no automation, anything that could help to clarify this matter.

> There are no approving reviews of non-core members either which would certainly help to push this.

I haven't seen neither any previous indication that approvals from non-core members count towards anything, nor a precedent that could cue me in on that, so either you just tried to manipulate this conversation, pointing at things that could have been done but weren't, or you've forgot to tell the community members it actually has any value...

> But I also don't think the tone of @didactic-drunk's previous comment is appropriate. And sadly I'm sure that such an attutide doesn't help to encourage anyone to do reviews.

What **is** and what **is not** appropriate is a very subjective matter. I for one, had such feeling many times when confronted by @RX14 simply rude behaviour - way off and below anything _appropriate_, especially for someone who IMO should steward this community, instead of treating it as its own playground. I wouldn't call stalling a PR an _appropriate_ either.

Reply:
Regarding the documentation:
1. I didn't change the documentation.  @straight-shoota did as an addition I assumed was necessary to get the PR approved.  Proof provided below.

What I was trying to say:
1.  People are requesting (documentation or other cleanup) changes which creates more requests for changes leaving this PR in perpetual limbo.
2. I wanted a quick code cleanup of the parts I'm interested in.  At what point can I say "not it" and let someone else handle a documentation cleanup without my involvement?
3. Without sarcasm or frustration: What will get this PR through faster?

Regarding tone:
1. Tone is hard to get.  I admit being annoyed initially, but tried to edit it out the emotional cruft and get down to a single purpose.  What gets the PR through faster?
2. The purpose of the last message was to provide minimal context highlighting the problems I see.  Others have mentioned I both write too much or too little.  Apparently my attempt to hit the middle came off wrong.
3. If there's something out of line in my post I don't see it.  I was calm while editing and removed what I saw as emotional.  The only parts left are what I think are relevant context.  Maybe it's too terse?  IDK.

```
Author: didactic-drunk <1479616+didactic-drunk@users.noreply.github.com>
Date:   Mon Nov 25 15:29:48 2019 -0800

    Update src/digest/base.cr

    Co-Authored-By: Johannes Müller <johannes.mueller@smj-fulda.org>

diff --git a/src/digest/base.cr b/src/digest/base.cr
index b321d47bc..bf573a687 100644
--- a/src/digest/base.cr
+++ b/src/digest/base.cr
@@ -14,7 +14,7 @@ abstract class Digest::Base
       end
     end

-    # Yields a context object with an `#update(data : String | Bytes)`
+    # Yields an instance of `self` which can receive calls to `#update(data : String | Bytes)` and returns the finalized digest.
     # method available. Returns the resulting digest afterwards.
     #
```

Reply:
If it's stlll not clear:
1. I made minimal code changes.
2. @straight-shoota wanted documentation/various cleanup changes (Mostly last year).
3. @RX14 is correcting @straight-shootas's (as I see it unrelated to my changes) changes.
4. Feedback loop noticed.  Why can't they fix the documentation in a separate PR?

^ Not bitchy.  Factual tone.

Reply:
> It would be way easier if we could have any estimates in re: to when can we expect any reaction

That would be really great, but I have no idea how it could be possible.

> who's responsible for doing so

We have no formalized code ownership or something similar, so it's just the whole core team in total responsible. I.e. nobody in particular. 

> what can we do to ease the process as much as possible (thinking of you)

Reviews are easy if they consist of only a few self-contained commits and don't change much during the PR discussion. That means the specification of the API should be finalized before the PR even starts and the PR itself should only be about implementation details.
So this PR is actually a very good example of a very hard review process. Over time it has collected a convoluted commit history and lots of comments. @didactic-drunk, please take no offense! It's just an observation from my point of view. Maybe it would have been better to put  the PR on hold right at the beginning and start a separate design discussion first.

> I haven't seen neither any previous indication that approvals from non-core members count towards anything

Honestly, I didn't expect there to be any doubt about that. It is true that PRs can only be merged with two core member approvals, but that doesn't mean non-core member's reviews are useless. I (and I can probably speak for the rest of the core team) am very happy about reviews from everyone in the community! And you're doing a great job at that. But the effective results are often only comments pointing out issues and suggesting improvements. When you do a review and think the changes are fine, please also leave an approval! It helps a lot to see such feedback from the community. I don't think this should be only a core team job.

Reply:
@didactic-drunk Sry, I didn't notice your comments prior to my previous one. As mentioned there the history of this PR is really a mess.

I guess misinterpreted the tone in your message. It sounded a bit bitchy to me. Sorry about that.

Reply:
Sorry, I didn't realise the documentation changes were applied from @straight-shoota.

It's really hard to review PRs with complex, long histories such as this one, and everyone gets it wrong. Let's just accept everyone makes mistakes and try to interpret other people's actions in good faith, as I try my best to do.

Reply:
I'm not holding a grudge.  My tone is and has been flat or questioning this entire post.  If only github had some sort of [tone or musical markup](https://github.com/watzon/nacl/issues/1#issuecomment-504715065).  Please reference the link for my usual tone when questioning my posts.

Reply:
My first 3 commits were split in an attempt to make the review process faster/easier.  Did that work or should I have a single jumbo commit when posting PR's?